File: csharp/exercise/exer01b-max-div.cs
/*
 * MAXIMUM NUMBER OF DIVISORS
 */
using System;
using System.Collections.Generic;
using System.Threading;
using System.Linq;



class Exer01B : IRunnable
{
    public void run()
    {
        const int RANGE_START = 1;
        const int RANGE_STOP = 100000;
        const int NUM_THREADS = 8;

        var lstWorkerArg = prepareArg(RANGE_START, RANGE_STOP, NUM_THREADS);
        var lstWorkerRes = new List<WorkerResult>();
        var lstTh = new List<Thread>();


        for (int ith = 0; ith < NUM_THREADS; ++ith)
        {
            var arg = lstWorkerArg[ith];

            lstTh.Add(new Thread(() =>
            {
                int resValue = 0;
                int resNumDiv = 0;

                for (int i = arg.iStart; i <= arg.iEnd; ++i)
                {
                    int numDiv = 0;

                    for (int j = i / 2; j > 0; --j)
                        if (i % j == 0)
                            ++numDiv;

                    if (resNumDiv < numDiv)
                    {
                        resNumDiv = numDiv;
                        resValue = i;
                    }
                }

                lock (lstWorkerRes)
                {
                    lstWorkerRes.Add(new WorkerResult(resValue, resNumDiv));
                }
            }));
        }


        var tpStart = DateTime.Now;

        lstTh.ForEach(th => th.Start());
        lstTh.ForEach(th => th.Join());
        var finalRes = lstWorkerRes.OrderByDescending(res => res.numDiv).First();

        var timeElapsed = (DateTime.Now - tpStart).TotalSeconds;

        Console.WriteLine("The integer which has largest number of divisors is " + finalRes.value);
        Console.WriteLine("The largest number of divisor is " + finalRes.numDiv);
        Console.WriteLine("Time elapsed = " + timeElapsed);


        /*
         * BETTER WAY (avoiding synchronization of lstWorkerRes):
         *
         * - Initialize lstWorkerRes with null objects.
         *   Of course, the number of objects is NUM_THREADS.
         *
         * - In thread function:
         *   lstWorkerRes[threadIndex] = new WorkerResult(resValue, resNumDiv);
         */
    }


    private List<WorkerArg> prepareArg(int rangeStart, int rangeEnd, int numThreads)
    {
        int rangeA, rangeB, rangeBlock;

        rangeBlock = (rangeEnd - rangeStart + 1) / numThreads;
        rangeA = rangeStart;

        var lstWorkerArg = new List<WorkerArg>();

        for (int i = 0; i < numThreads; ++i, rangeA += rangeBlock)
        {
            rangeB = rangeA + rangeBlock - 1;

            if (i == numThreads - 1)
                rangeB = rangeEnd;

            lstWorkerArg.Add(new WorkerArg(rangeA, rangeB));
        }

        return lstWorkerArg;
    }


    readonly struct WorkerArg
    {
        public readonly int iStart, iEnd;
        public WorkerArg(int iStart, int iEnd) => (this.iStart, this.iEnd) = (iStart, iEnd);
    }


    readonly struct WorkerResult
    {
        public readonly int value, numDiv;
        public WorkerResult(int value, int numDiv) => (this.value, this.numDiv) = (value, numDiv);
    }
}

File: python/exer01b_max_div.py
'''
MAXIMUM NUMBER OF DIVISORS
'''

import time
import threading



lk = threading.Lock()



def prepare_arg(rng_start: int, rng_end: int, num_threads: int) -> list:
    rng_block = (rng_end - rng_start + 1) // num_threads
    rng_a = rng_start
    lst_arg = []

    for i in range(num_threads):
        rng_b = rng_a + rng_block - 1 if i < num_threads - 1 else rng_end
        lst_arg.append({ 'start': rng_a, 'end': rng_b })
        rng_a += rng_block

    return lst_arg



def do_task(arg, lst_res: list):
    res_value = 0
    res_numdiv = 0

    for i in range(arg['start'], arg['end'] + 1):
        numdiv = 0

        for j in range(1, i // 2):
            if i % j == 0:
                numdiv += 1

        if res_numdiv < numdiv:
            res_numdiv = numdiv
            res_value = i

    with lk:
        lst_res.append({ 'value': res_value, 'numdiv': res_numdiv })

    '''
    BETTER WAY (avoiding synchronization of lst_res):

    - Initialize lst_res with null objects.
        Of course, the number of objects is NUM_THREADS.

    - In thread function:
        lst_res[thread_index] = { 'value': res_value, 'numdiv': res_numdiv }
    '''



##########################################################



RANGE_START = 1
RANGE_END = 30000
NUM_THREADS = 8

lst_worker_arg = prepare_arg(RANGE_START, RANGE_END, NUM_THREADS)
lst_worker_res = []

lstth = [
    threading.Thread( target=do_task, args=(arg, lst_worker_res) )
    for arg in lst_worker_arg
]


tp_start = time.time()

for th in lstth:
    th.start()

for th in lstth:
    th.join()

final_res = sorted(lst_worker_res, key=lambda res: res['numdiv'])[-1]

time_elapsed = time.time() - tp_start


print('The integer which has largest number of divisors is', final_res['value'])
print('The largest number of divisor is', final_res['numdiv'])
print('Time elapsed =', time_elapsed)
